packages/flutter_tools/lib/src/test/integration_test_device.dart

```diff
@@ -0,0 +1,140 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+// @dart = 2.8
+
+import 'dart:async';
+
+import 'package:meta/meta.dart';
+import 'package:stream_channel/stream_channel.dart';
+import 'package:vm_service/vm_service.dart' as vm_service;
+
+import '../application_package.dart';
+import '../base/common.dart';
+import '../build_info.dart';
+import '../device.dart';
+import '../globals.dart' as globals;
+import '../vmservice.dart';
+import 'test_device.dart';
+
+const String kIntegrationTestExtension = 'Flutter.IntegrationTest';
+const String kIntegrationTestData = 'data';
+const String kIntegrationTestMethod = 'ext.flutter.integrationTest';
+
+class IntegrationTestTestDevice implements TestDevice {
+  IntegrationTestTestDevice({
+    @required this.id,
+    @required this.device,
+    @required this.debuggingOptions,
+    @required this.userIdentifier,
+  });
+
+  final int id;
+  final Device device;
+  final DebuggingOptions debuggingOptions;
+  final String userIdentifier;
+
+  ApplicationPackage _applicationPackage;
+  final Completer<void> _finished = Completer<void>();
+  final Completer<Uri> _gotProcessObservatoryUri = Completer<Uri>();
+
+  /// Starts the device.
+  ///
+  /// [entrypointPath] must be a path to an uncompiled source file.
+  @override
+  Future<StreamChannel<String>> start(String entrypointPath) async {
+    final TargetPlatform targetPlatform = await device.targetPlatform;
+    _applicationPackage = await ApplicationPackageFactory.instance.getPackageForPlatform(
+      targetPlatform,
+      buildInfo: debuggingOptions.buildInfo,
+    );
+
+    final LaunchResult launchResult = await device.startApp(
+      _applicationPackage,
+      mainPath: entrypointPath,
+      debuggingOptions: debuggingOptions,
+      userIdentifier: userIdentifier,
+    );
+    assert(launchResult.started);
+
+    // No need to set up the log reader because the logs are captured and
+    // streamed to the package:test_core runner.
+
+    _gotProcessObservatoryUri.complete(launchResult.observatoryUri);
+
+    globals.printTrace('test $id: Connecting to vm service');
+    final vm_service.VmService vmService = await connectToVmService(launchResult.observatoryUri);
+
+    globals.printTrace('test $id: Finding the correct isolate with the integration test service extension');
+    final vm_service.IsolateRef isolateRef = await _findExtensionIsolate(vmService);
```

Comment:
There may be some timing issues here, since the kIntegrationTestMethod will not be registered until after main finishes, but this may be after the app/isolate has started. There is a waitForExtension method in the resident_runner library that handles this cases, though it doesn't identify the correct Isolate:

https://github.com/flutter/flutter/blob/dea1c46a3c046766089a5d2e85bf6e169411b82e/packages/flutter_tools/lib/src/resident_devtools_handler.dart#L186

Reply:
Would there be issues with the implementation of `_findExtensionIsolate` method defined below? It looks like it's implementing something similar to the `waitForExtension` method (both subscribing to registered extension events as well as querying for current extensions, which should solve the timing issues here

Reply:
Yeah that makes sense - I would prefer trying to reuse the existing helper if possible.

Reply:
Refactoring to make the existing helper reusable here was done in #77220, and just updated this PR to use the helper here.